proxy/src/main/java/com/velocitypowered/proxy/connection/registry/DimensionRegistry.java

```diff
@@ -0,0 +1,109 @@
+package com.velocitypowered.proxy.connection.registry;
+
+import com.google.common.base.Preconditions;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.collect.Maps;
+
+import java.util.Map;
+import java.util.Set;
+
+import net.kyori.nbt.CompoundTag;
+import net.kyori.nbt.ListTag;
+import net.kyori.nbt.Tag;
+import net.kyori.nbt.TagType;
+import org.checkerframework.checker.nullness.qual.Nullable;
+
+
+public final class DimensionRegistry {
+
+  private final Map<String, DimensionData> registeredDimensions;
+  private final ImmutableSet<String> levelNames;
+
+  /**
+   * Initializes a new {@link DimensionRegistry} instance.
+   * This registry is required for 1.16+ clients/servers to communicate,
+   * it constrains the dimension types and names the client can be sent
+   * in a Respawn action (dimension change).
+   * This WILL raise an IllegalArgumentException if the following is not met:
+   * - At least one valid DimensionData instance is provided
+   * - At least one valid world name is provided
+   * @param registeredDimensions a populated {@link ImmutableSet} containing dimension data types
+   * @param levelNames a populated {@link ImmutableSet} of the level (world) names the server offers
+   */
+  public DimensionRegistry(ImmutableSet<DimensionData> registeredDimensions,
+                            ImmutableSet<String> levelNames) {
+    Preconditions.checkNotNull(registeredDimensions,
+            "registeredDimensions cannot be null");
+    Preconditions.checkNotNull(levelNames,
+            "levelNames cannot be null");
+    Preconditions.checkArgument(registeredDimensions.size() > 0,
+            "registeredDimensions needs to be populated");
+    Preconditions.checkArgument(levelNames.size() > 0,
+            "levelNames needs to populated");
+    this.registeredDimensions = Maps.uniqueIndex(
+            registeredDimensions, DimensionData::getRegistryIdentifier);
+    this.levelNames = levelNames;
+  }
+
+  public Map<String, DimensionData> getRegisteredDimensions() {
+    return registeredDimensions;
+  }
+
+  public Set<String> getLevelNames() {
+    return levelNames;
+  }
+
+  /**
+   * Returns the internal dimension data type as used by the game.
+   * @param dimensionIdentifier how the dimension is identified by the connection
+   * @return game dimension data or null if not registered
+   */
+  public @Nullable DimensionData getDimensionData(String dimensionIdentifier) {
+    return registeredDimensions.getOrDefault(dimensionIdentifier, null);
```

Comment:
Can be replaced with `registeredDimensions.get(dimensionIdentifier)`

Reply:
Ah I just read up that it returns *null* in java for a map key that does not have a value and does NOT throw an exception instead, my bad, will fix.